wallet/src/main/scala/org/bitcoins/wallet/internal/AddressHandling.scala

```diff
@@ -206,4 +144,116 @@ private[wallet] trait AddressHandling extends WalletLogger {
     }
   }
 
+  /** Background thread meant to ensure safety when calling [[getNewAddress()]]
+    * We to ensure independent calls to getNewAddress don't result in a race condition
+    * to the database that would generate the same address and cause an error.
+    * With this background thread, we poll the [[addressRequestQueue]] seeing if there
+    * are any elements in it, if there are, we process them and complete the Promise in the queue. */
+  lazy val walletThread = new Thread(AddressQueueRunnable)
+
+  lazy val addressRequestQueue = {
+    new java.util.concurrent.ArrayBlockingQueue[(AccountDb, HDChainType, Promise[BitcoinAddress])](
+      walletConfig.addressQueueSize
+    )
+  }
+
+  walletThread.setDaemon(true)
+  walletThread.setName(
+    s"wallet-address-queue-${System.currentTimeMillis()}")
+  walletThread.start()
+  
+  /** A runnable that drains [[addressRequestQueue]]. Currently polls every 100ms
+    * seeing if things are in the queue. This is needed because otherwise
+    * wallet address generation is not async safe.
+    * @see https://github.com/bitcoin-s/bitcoin-s/issues/1009
+    * */
+  private case object AddressQueueRunnable extends Runnable {
```

Comment:
On the `dlc` branch I have split up the former function `getNewAddressHelper` into 2 separate functions, one that gets the addressdb and one that writes. Would that break here / is it possible to do that?

Reply:
So I think you should maybe bring those changes upstream so I can incorporate them here: 

To answer your question, I believe the writing is the most important -- as long as the read is not used to create a new address. The errors that were caused before this PR was because there was a race condition when writing to the database.

Reply:
https://github.com/bitcoin-s/bitcoin-s/pull/1322